### ISSUE TITLE
Propagate backpressure by not accepting new connections for TCP-based sources

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#8596]
We observed in https://github.com/timberio/vector/issues/8238 that, in the presence of back-pressure, Vector does not stop accepting new TCP connections for TCP-based sources. Instead it accepts them and deserializes the first "message" from the wire, blocking when it attempts to push to the next components (`Pipeline.send`). This will cause it to hold open connections and memory.

Similar to https://github.com/timberio/vector/issues/7186 I think we should have a configurable ceiling after which Vector will refuse to accept additional connections while it processes those in-flight.

**Comment by anas-aso** [comment]
Hi @jszwedko. I am affected by this issue. I have noticed that Vector doesn't close the connection at all in some cases and this leads to a high memory usage even when the incoming events rate becomes low.
I was thinking about setting the keepalive to mitigate the connections leak (https://vector.dev/docs/reference/configuration/sources/fluent/#keepalive and https://docs.fluentd.org/output/forward#keepalive). Do you think this might help ?

**Comment by jszwedko** [comment]
Hi @anas-aso ! Do you mean you want Vector to close idle connections?

**Comment by Cbeck527** [comment]
I wanted to add some more info here in case it was helpful. We're also seeing this when processing _a lot_ (to us) of log events with the Vector source.

To spare you the boring in-depth architecture details: we have a fleet of Vector boxes behind an NLB that does all of our log processing. Our clients are wired up with the Vector sink to send to this AWS NLB-fronted DNS address. The log processing boxes do all the heavy lifting and then dump to an elasticsearch cluster on the other end.

What's particularly interesting is we saw servers in our autoscaling group start to fall over when upgrading from 0.18.1 to 0.20.0. I didn't have enough time to gather more details (we rolled back) but what I _did_ see is we hit our open file limit for the vector process.

Something like 

```
lsof -p $VECTOR_PID | grep CLOSE_WAIT
```

was super high. I searched `ulimit` in issues to see if anyone had run into something similar, and ended up getting linked to https://github.com/vectordotdev/vector/issues/8238 — and then here.

I hope this helps! More than happy to answer any additional questions. I'm going to attempt to upgrade again in our lab environment to see what happens, but I most likely won't get to that for about a week or so.
